File: templates/problems/09_heap/01_kth_largest_stream/solution.go
package kth_largest_stream

import "container/heap"

// SolveKthLargest uses a min-heap of size k to track the kth largest element.
// Constructor: O(n log k), Add: O(log k), Space: O(k).
type SolveKthLargest struct {
	k    int
	minH *solveMinHeap
}

type solveMinHeap []int

func (h solveMinHeap) Len() int            { return len(h) }
func (h solveMinHeap) Less(i, j int) bool   { return h[i] < h[j] }
func (h solveMinHeap) Swap(i, j int)        { h[i], h[j] = h[j], h[i] }
func (h *solveMinHeap) Push(x interface{})  { *h = append(*h, x.(int)) }
func (h *solveMinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func SolveConstructor(k int, nums []int) SolveKthLargest {
	h := &solveMinHeap{}
	heap.Init(h)
	kl := SolveKthLargest{k: k, minH: h}
	for _, n := range nums {
		kl.SolveAdd(n)
	}
	return kl
}

func (kl *SolveKthLargest) SolveAdd(val int) int {
	heap.Push(kl.minH, val)
	if kl.minH.Len() > kl.k {
		heap.Pop(kl.minH)
	}
	return (*kl.minH)[0]
}

File: templates/problems/09_heap/01_kth_largest_stream/python/solution.py
import heapq


class SolveKthLargest:
    """Track the kth largest element in a stream using a min-heap of size k.

    Constructor: O(n log k), Add: O(log k), Space: O(k).
    """

    def __init__(self, k: int, nums: list[int]) -> None:
        self.k = k
        self._heap: list[int] = []
        for n in nums:
            self.solve_add(n)

    def solve_add(self, val: int) -> int:
        heapq.heappush(self._heap, val)
        if len(self._heap) > self.k:
            heapq.heappop(self._heap)
        return self._heap[0]

    # Alias matching the stub method name for convenience.
    def add(self, val: int) -> int:
        return self.solve_add(val)
